### ISSUE TITLE
[Feature]: add automatic generation of password secrets for managed roles

### ISSUE THREAD

**Issue by jsilvela** [cloudnative-pg/cloudnative-pg#3788]
### Is there an existing issue already for this feature request/idea?

- [X] I have searched for an existing issue, and could not find anything. I believe this is a new feature request to be evaluated.

### What problem is this feature going to solve? Why should it be added?

The current implementation of managed roles allows giving the name
of a secret where a role's password will be kept, but that secret is
assumed to exist, and its creation falls o the user.

For users migrating from other solutions, or users wanting to get on
CloudNativePG and have password-protected roles from the get-go,
it would be convenient if the secrets were generated by CloudNativePG
as needed.

See thread https://cloudnativepg.slack.com/archives/C03AX0J5P29/p1706780445388369

### Describe the solution you'd like

Allow the managed roles stanza to specify either an existing secret
to hold a role's password, or an auto-generated secret.

### Describe alternatives you've considered

at the moment, automatic creation of password secrets would
fall outside of the CNPG envelope

### Backport?

Yes

### Are you willing to actively contribute to this feature?

Yes

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by gpothier** [comment]
I agree that it would be a nice thing to have. Additionally, and maybe this should be filed in a separate issue, if one sets a value for `passwordSecret.name` and that secret does not exist, the cluster does not complain at all; I think this is not correct. I tried to set a random password secret name just to see if the secret would be created if it does not exist. IMHO either the cluster creation should fail (or at least report a warning), or the secret should be created. Although I think it would be better to explicitly require the secret to be created.

**Comment by isdinAghmir** [comment]
Hello guys ! 

is this a good candidate for the roadmap ? 

We are currently facing the same issue.

**Comment by jeremych1000** [comment]
This would be great for us - we get CNPG to create the roles for us and autogenerating passwords would reduce plumbing by a LOT.

**Comment by szechyjs** [comment]
It would be great if this worked just like the auto generated `app` role, which creates a secret that contains connection details in addition to credentials.

**Comment by hopkinsjason** [comment]
Yes, this would be a great feature - I'm migrating from CrunchyData's PGO and this is causing me pain!  PGO has the concept of creating multiple databases/users in a cluster and for each user it creates a secret very similar to the app secret create by CNPG.

A single app user works great if you can accept the management overhead/cost of a db per service, but that's not always desirable or required.

**Comment by longtomjr** [comment]
I think it is still worth keeping this ticket open (and not stale). This would be valuable quality of life for provisioning and migrating from other operators.

**Comment by yevon** [comment]
This would be great allowing to auto manage passwords and create secrets as its done with the superuser.

**Comment by fr6nco** [comment]
anyone struggling with this, Im using this [operator](https://github.com/mittwald/kubernetes-secret-generator) to generate passwords:

```yaml
---
apiVersion: "secretgenerator.mittwald.de/v1alpha1"
kind: "BasicAuth"
metadata:
  name: secretname
spec:
  forceRegenerate: false
  username: username
  length: "32"
```

works well with cnpg, on cnpg cluster then I set:
```yaml
    roles:
      - name: username
        ensure: present
        login: true
        superuser: true
        comment: "Vault identity for dynamic credentials"
        passwordSecret:
          name: secretname
```

as I result I get a secret with username / password combination

**Comment by YvesEarnix** [comment]
Additionally should great to **support password rotation**, it's will be facilitate user that want to migration from Zalando operator to cloudnative-pg

https://opensource.zalando.com/postgres-operator/docs/reference/cluster_manifest.html
**usersWithSecretRotation** list of users to enable credential rotation in K8s secrets. The rotation interval can only be configured globally. On each rotation a new user will be added in the database replacing the username value in the secret of the listed user. Although, rotation users inherit all rights from the original role, keep in mind that ownership is not transferred. See more details in the [administrator docs](https://github.com/zalando/postgres-operator/blob/master/docs/administrator.md#password-rotation-in-k8s-secrets).

**usersWithInPlaceSecretRotation** list of users to enable in-place password rotation in K8s secrets. The rotation interval can only be configured globally. On each rotation the password value will be replaced in the secrets which the operator reflects in the database, too. List only users here that rarely connect to the database, like a flyway user running a migration on Pod start. See more details in the [administrator docs](https://github.com/zalando/postgres-operator/blob/master/docs/administrator.md#password-replacement-without-extra-users).

**usersIgnoringSecretRotation** if you have secret rotation enabled globally you can define a list of of users that should opt out from it, for example if you store credentials outside of K8s, too, and corresponding deployments cannot dynamically reference secrets. Note, you can also opt out from the rotation by removing users from the manifest’s users section. The operator will not drop them from the database. Optional.

https://opensource.zalando.com/postgres-operator/docs/reference/operator_parameters.html#postgres-users

**enable_password_rotation** For all LOGIN roles that are not database owners the operator can rotate credentials in the corresponding K8s secrets by replacing the username and password. This means, new users will be added on each rotation inheriting all priviliges from the original roles. The rotation date (in YYMMDD format) is appended to the names of the new user. The timestamp of the next rotation is written to the secret. The default is false.

**password_rotation_interval** If password rotation is enabled (either from config or cluster manifest) the interval can be configured with this parameter. The measure is in days which means daily rotation (1) is the most frequent interval possible. Default is 90.

**password_rotation_user_retention** To avoid an ever growing amount of new users due to password rotation the operator will remove the created users again after a certain amount of days has passed. The number can be configured with this parameter. However, the operator will check that the retention policy is at least twice as long as the rotation interval and update to this minimum in case it is not. Default is 180.

**Comment by jsilvela** [comment]
Should probably not be done before #5341

**Comment by mjsir911** [comment]
Might be related, can't find a more relevant issue to this, but somehow I would like to be able to get a secret with a postgresql `uri` key for a `Deployment` to use within it's env, even when the `Cluster` is being defined in a different namespace. I'm defining my `Database` (& in the future, `Role`) within my application's namespace

**Comment by Untersander** [comment]
Would love this feature. I will stay with crunchy pgo and give cnpg another shot if this feature is implemented.

**Comment by Edward13ruf** [comment]
I also am blocked on migrating to cnpg because of this, as I suspect many people are. Generation and automatic password rotation are critical to our security workload. Ideally we'd like to generate those secrets in other namespaces, but regardless we want to reduce any manual maintenance from password management in our cluster.

**Comment by jayantasamaddar** [comment]
I wrote a custom Makefile script to make it work. Sharing my setup.:

## In a `.env` declare:

```
APP_NAMESPACE=yournamespace
PG_PORT=5432
PG_DATABASE=mydb
## other env vars
```

## In Makefile:

```Makefile
# Include your environment variables
include .env

# Define a function for reusability. Pass an username to generate a password.
define create-cnpg-user-secret
USERNAME=$(1); \
PASS=$$(openssl rand -hex 32 | tr -d '\n'); \
HOST="$(APP_NAMESPACE)-postgres-rw.$(APP_NAMESPACE)"; \
PORT="$(PG_PORT)"; \
URI="postgresql://$$USERNAME:$$PASS@$$HOST:$$PORT/$(PG_DATABASE)"; \
FQDN_URI="postgresql://$$USERNAME:$$PASS@$$HOST:$$PORT/$(PG_DATABASE)"; \
JDBC_URI="jdbc:postgresql://$$HOST.svc.cluster.local:$$PORT/$(PG_DATABASE)?password=$$PASS&user=$$USERNAME"; \
FQDN_JDBC_URI="jdbc:postgresql://$$HOST.svc.cluster.local:$$PORT/$(PG_DATABASE)?password=$$PASS&user=$$USERNAME"; \
PGPASS="$(APP_NAMESPACE)-postgres-rw:$$PORT:$(PG_DATABASE):$$USERNAME:$$PASS"; \
printf "%s\n" \
"apiVersion: v1" \
"kind: Secret" \
"metadata:" \
"  name: $(APP_NAMESPACE)-postgres-$$USERNAME" \
"  namespace: $(APP_NAMESPACE)" \
"  labels:" \
"    cnpg.io/reload: \"true\"" \
"type: kubernetes.io/basic-auth" \
"stringData:" \
"  user: $$USERNAME" \
"  username: $$USERNAME" \
"  password: $$PASS" \
"  pgpass: $$PGPASS" \
"  host: $$HOST" \
"  port: \"$$PORT\"" \
"  dbname: $(PG_DATABASE)" \
"  uri: $$URI" \
"  fqdn-uri: $$FQDN_URI" \
"  jdbc-uri: $$JDBC_URI" \
"  fqdn-jdbc-uri: $$FQDN_JDBC_URI" \
| kubectl create -f -
endef


# Commands

# Create multiple postgres roles and generate password:
## Syntax: pg-role <username1> <username2>
pg-role:
	@users="$(filter-out $@,$(MAKECMDGOALS))"; \
	if [ -z "$$users" ]; then \
		echo "Usage: make pg-user <username> [username...]"; \
		exit 1; \
	fi; \
	for u in $$users; do \
		$(call create-cnpg-user-secret,$$u); \
	done

# Ignore extra args only when pg-user is the main goal
ifneq ($(filter pg-user,$(MAKECMDGOALS)),)
%:
	@:
endif

# Create a single postgres role and generate password:
## Syntax: pg-role-<username>
pg-role-%:
	@echo Creating role $*
	$(call create-cnpg-user-secret,$*)
```

---

## Invocation

```Makefile
# Use like so:

## Syntax 1: Multiple
make pg-role newuser1 newuser2     # creates two roles with generated passwords

## Syntax 2: make pg-user-<username>
make pg-role-nexthis-newuser -s       # creates a newuser role with a generated password
```

---

## Deployment of Infra


### Cluster Setup

```yaml
apiVersion: postgresql.cnpg.io/v1
kind: Cluster
metadata:
  name: nexthis-global-postgres
spec:
  instances: 2 # 1 RW + 1 Replica
  # imageName: ghcr.io/cloudnative-pg/postgresql:17.4.5-bookworm # Use PostgreSQL 17. Specify when you want to directly set image repository
  imageCatalogRef:
    apiGroup: postgresql.cnpg.io
    kind: ClusterImageCatalog
    name: postgresql-standard-trixie
    major: 18
  storage:
    size: 1Gi # 1GB per instance
  enableSuperuserAccess: true # Enable superuser access

  # Bootstrap database settings
  bootstrap:
    initdb:
      database: global # Set custom database name
      owner: nexthis # Optional: Set a custom

  managed:
    roles:
       # mention your custom roles here
      - name: nexthis-admin
        comment: Performs database migrations and owns all the tables and schema
        login: true
        passwordSecret:
          name: nexthis-global-postgres-nexthis-admin
```

### Makefile commands

```Makefile
pg:
# 	Create Namespace if it doesn't exist
	kubectl create ns $(APP_NAMESPACE) --dry-run=client -o yaml | kubectl apply -f -
	make pg-role-nexthis-admin -s # Creation of the additional user
	kubectl apply -n $(APP_NAMESPACE) -f ./kubernetes/pg.yml
	make pg-pooler -s
	make pg-publication -s

clean-pg:
	make clean-pg-publication -s
	make clean-pg-pooler -s
	kubectl delete -n $(APP_NAMESPACE) -f ./kubernetes/pg.yml
	kubectl delete ns $(APP_NAMESPACE)

pg-publication:
	kubectl apply -n $(APP_NAMESPACE) -f ./kubernetes/pg-publication.yml

clean-pg-publication:
	kubectl delete -n $(APP_NAMESPACE) -f ./kubernetes/pg-publication.yml

pg-pooler:
	kubectl apply -n $(APP_NAMESPACE) -f ./kubernetes/pg-pooler.yml

clean-pg-pooler:
	kubectl delete -n $(APP_NAMESPACE) -f ./kubernetes/pg-pooler.yml
```

### Deploy and Undeploy

```Makefile
# Deploy
make pg

# Undeploy
make clean-pg
```

---

> **Note**: If you create roles after the cluster is deployed, simply add it to the `cluster.spec.managed.roles` list and CloudNativePG creates the role in the database.
> For removal of a role, you have to -
> - Remove it from the `cluster.spec.managed.roles` list
> - Remove it from the database level and optionally
> - (Optionally) Clean the secret

> **P.S.**: I have not tested, if `managed` role means, password modification in the secret also modifies database password.

**Comment by gbartolini** [comment]
I know it is not the same thing, but version 1.30.0 introduces automated generation of client TLS certificates.

Regarding this feature: could not this be solved using the External Secrets Operator (ESO) as explained at https://cloudnative-pg.io/docs/current/cncf-projects/external-secrets?

**Comment by 9numbernine9** [comment]
> Regarding this feature: could not this be solved using the External Secrets Operator (ESO) as explained at https://cloudnative-pg.io/docs/current/cncf-projects/external-secrets?

_Personally_, I would rather use ESO to synchronize the passwords that are generated by CNPG to some other external secret store, rather than the opposite direction. In my ideal world, I'd love to see CNPG create the passwords for the new managed roles and use ESO to sync those into our external secret store.

**Comment by jeremych1000** [comment]
> > Regarding this feature: could not this be solved using the External Secrets Operator (ESO) as explained at https://cloudnative-pg.io/docs/current/cncf-projects/external-secrets?
> 
> _Personally_, I would rather use ESO to synchronize the passwords that are generated by CNPG to some other external secret store, rather than the opposite direction. In my ideal world, I'd love to see CNPG create the passwords for the new managed roles and use ESO to sync those into our external secret store.

Agreed - also ESO is yet another operator to maintain / write docs for / update for a platform engineering team. Most db operators have some sort of password generation so I think of this more of a feature parity request.

**Comment by gbartolini** [comment]
I have been thinking of adding a `password` stanza to `DatabaseRole` which should mimic what we do for the `clientCertificate` section introduced in #10809 (improved with #11312). We should also look at the fields already available in https://external-secrets.io/main/api/generator/password/ and adopt a similar [interface](https://github.com/external-secrets/external-secrets/blob/main/apis/generators/v1alpha1/types_password.go#L24). It could look something like this:

```
apiVersion: postgresql.cnpg.io/v1
kind: DatabaseRole
metadata:
  name: role-dante
spec:
  cluster:
    name: cluster-example
  name: dante
  ... <snip> ...
  password:
    enabled: true
    duration: 2160h      # 90 days, defaults to the operator's CertificateDuration setting
    renewBefore: 168h     # 7 days, defaults to the operator's ExpiringCheckThreshold setting
    criteria: # see https://github.com/external-secrets/external-secrets/blob/main/apis/generators/v1alpha1/types_password.go#L24
      length: ...
      symbolCharacters: ...
      ... <snip> ...
    secret: cluster-example-dante
  #passwordSecret:                         DEPRECATED/OBSOLETED
    #name: cluster-example-dante
```

This will be mutually exclusive with `passwordSecret` and potentially obsolete it.

What do you think?

**Comment by gbartolini** [comment]
FYI, I have submitted this patch to the PostgreSQL project for consideration in version 20: https://www.postgresql.org/message-id/CA%2BVUV5pyUB-sXUCxnOATJdE_XDdos6rAgW5wPVTz-0AT8bLuGw%40mail.gmail.com. Title: *Tracking role modification timestamps in pg_authid / pg_roles*.
